src/Abc.Zebus/SubscriptionHandling/SubscriptionSnapshotGenerator.cs

```diff
@@ -0,0 +1,38 @@
+using System;
+using Abc.Zebus.Directory;
+
+namespace Abc.Zebus.SubscriptionHandling
+{
+    /// <summary>
+    /// Extend this class to generate snapshots each time a new subscription is created for <see cref="TMessage"/>
+    /// </summary>
+    /// <typeparam name="TSnapshotMessage">The type of the snapshot message</typeparam>
+    /// <typeparam name="TMessage">The type of the message whose subscriptions will trigger a snapshot</typeparam>
+    public abstract class SubscriptionSnapshotGenerator<TSnapshotMessage, TMessage> : SubscriptionHandler<TMessage>
+        where TSnapshotMessage : IEvent
+        where TMessage : IEvent
+    {
+        private readonly IBus _bus;
+
+        protected SubscriptionSnapshotGenerator(IBus bus)
+        {
+            _bus = bus;
+        }
+
+        protected override void OnSubscription(SubscriptionsForType subscription, PeerId peerId)
+        {
+            var snapshot = GenerateSnapshot(subscription);
+            if (_bus is IInternalBus internalBus)
+                internalBus.Publish(snapshot, peerId);
+            else
+                throw new Exception("The bus is not an internal bus");
```

Comment:
- Don't use `Exception` but a subtype
- *"internal bus"* does not mean anything to the user, the error message needs to be better

Reply:
As above, the user will never see this message.

Reply:
Same response as above 😉